src/pineko/kfactor.py

```diff
@@ -209,148 +193,108 @@ def create_singlegridonly(
-        order_exists=order_exists,
-    )
+    # the actual alpha_s order to update
+    order_to_update = pto_to_update + min_as - 1
 
+    # check if the order is already there
+    is_in = is_already_in((order_to_update, min_al, 0, 0), grid_orders_filtered)
 
-def do_it(
-    centrals_kfactor,
-    alphas,
-    grid_path,
-    grid,
-    max_as,
-    max_as_test,
-    target_folder,
-    order_exists,
-):
-    """Apply the centrals_kfactor to the grid if the order is not already there."""
-    grid_orders = [orde.as_tuple() for orde in grid.orders()]
-    order_mask = pineappl.grid.Order.create_mask(grid.orders(), max_as, 0, True)
-    grid_orders_filtered = list(np.array(grid_orders)[order_mask])
-    grid_orders_filtered.sort(key=scale_variations.qcd)
-    first_nonzero_order = grid_orders_filtered[0]
-    min_al = first_nonzero_order[1]
-    is_in = is_already_in(
-        (first_nonzero_order[0] + max_as_test, min_al, 0, 0), grid_orders_filtered
-    )
+    # Prevent summing orders incoherently
     if is_in and not order_exists:
         rich.print("[green] Success: Requested order already in the grid.")
         return
     if not is_in and order_exists:
         rich.print("[red] Abort: order exists is True but order not in the grid.")
         return
-    construct_and_merge_grids(
-        grid_path,
-        max_as_test,
-        first_nonzero_order,
-        min_al,
-        centrals_kfactor,
-        alphas,
-        target_folder,
-        order_exists,
-    )
+
+    # loop on all the order to update
+    max_as = grid_orders_filtered[-1][0]
+    orders_list = [(de, min_al, 0, 0) for de in range(min_as, max_as + 1)]
+    # create an empty grid and add the rescaled order
+    order_to_update = (order_to_update, min_al, 0, 0)
+    new_order_grid = None
+    for i, as_order in enumerate(orders_list):
+        order_grid = construct_new_order(
+            grid, as_order, order_to_update, central_kfactor, alphas
+        )
+        if i == 0:
+            new_order_grid = order_grid
+        else:
+            new_order_grid.merge(order_grid)
+
+    new_grid = grid
+    # if the new order is there, clean the old one.
+    if is_in:
+        new_grid = scale_variations.construct_and_dump_order_exists_grid(
+            grid, order_to_update
+        )
+    # merge the updated order with the original one.
+    new_grid.merge(new_order_grid)
+    new_grid.write_lz4(target_grid_path)
 
 
 def filter_k_factors(pigrid, centrals_kfactor):
-    """Filter the centrals k-factors according to their lenght compared to the number of bins of the grid."""
+    """Filter the centrals k-factors according to their length compared to the number of bins of the grid."""
     centrals_kfactor_filtered = np.array([])
     if pigrid.bins() == len(centrals_kfactor):
-        rich.print("[orange] The number of bins match the lenght of the k-factor.")
+        rich.print("[orange] The number of bins match the length of the k-factor.")
```

Comment:
```suggestion
        rich.print("[green] The number of bins match the length of the k-factor.")
```
this is good, right?

Reply:
I actually don't understand how this case can work:

https://github.com/NNPDF/pineko/blob/06458035efaa06e5236acdd07cffde12afa2bfe6/src/pineko/kfactor.py#L336-L342

but fine...

Reply:
What do you mean?

Reply:
In fact you can see that in line 339 I am checking that indeed the kfactor is a constant

Reply:
> Yes but the green is for when the job is done,

fine, then let's drop the color - orange sounds too much like a warning to me

Reply:
done in [55e015d](https://github.com/NNPDF/pineko/pull/161/commits/55e015d7f60753c0982d059cbfc6fb4c32c37aab)